adapter/src/test/java/com/twilio/kudu/sql/NestedJoinCacheTest.java

```diff
@@ -0,0 +1,115 @@
+import org.junit.Test;
+
+public final class NestedJoinCacheTest {
+  @Test
+  public void cacheTest() {
+    final List<TranslationPredicate> joinTranslations = Collections
+        .singletonList(new TranslationPredicate(1, 0, ComparisonOp.EQUAL));
+    final CloneableEnumerable<Object> mockBaseEnumerable = mock(CloneableEnumerable.class);
+    final NestedJoinFactory joinFactory = new NestedJoinFactory(10, joinTranslations, mockBaseEnumerable);
+    final CloneableEnumerable<Object> mockEnumerableFirstRow = mock(CloneableEnumerable.class);
+    final CloneableEnumerable<Object> mockEnumerableSecondRow = mock(CloneableEnumerable.class);
+    final CloneableEnumerable<Object> mockEnumerableThirdRow = mock(CloneableEnumerable.class);
+    final List<Object> firstBatch = Arrays.asList(new Object[] { "Awesome", Integer.valueOf(1) },
+        new Object[] { "Awesome2", Integer.valueOf(2) }, new Object[] { "Awesome3", Integer.valueOf(3) });
+
+    List<List<CalciteKuduPredicate>> firstRowPredicates = Collections
+        .singletonList(Arrays.asList(new ComparisonPredicate(0, ComparisonOp.EQUAL, Integer.valueOf(1))));
+    List<List<CalciteKuduPredicate>> secondRowPredicates = Collections
+        .singletonList(Arrays.asList(new ComparisonPredicate(0, ComparisonOp.EQUAL, Integer.valueOf(2))));
+    List<List<CalciteKuduPredicate>> thirdRowPredicates = Collections
+        .singletonList(Arrays.asList(new ComparisonPredicate(0, ComparisonOp.EQUAL, Integer.valueOf(3))));
+
+    when(mockBaseEnumerable.clone(firstRowPredicates)).thenReturn(mockEnumerableFirstRow);
+    when(mockBaseEnumerable.clone(secondRowPredicates)).thenReturn(mockEnumerableSecondRow);
+    when(mockBaseEnumerable.clone(thirdRowPredicates)).thenReturn(mockEnumerableThirdRow);
+
+    when(mockEnumerableFirstRow.enumerator()).thenReturn(
+        new ListEnumerator<Object>(Collections.singletonList(new Object[] { Integer.valueOf(1), "OneRow" })));
+    when(mockEnumerableSecondRow.enumerator()).thenReturn(
+        new ListEnumerator<Object>(Collections.singletonList(new Object[] { Integer.valueOf(2), "TwoRow" })));
+    when(mockEnumerableThirdRow.enumerator()).thenReturn(
+        new ListEnumerator<Object>(Collections.singletonList(new Object[] { Integer.valueOf(3), "ThirdRow" })));
+
+    // Run the same rows through a bunch of times.
+    for (int i = 0; i < 1000; i++) {
+      final List<Object> rightHandSide = joinFactory.apply(firstBatch).toList();
+
+      assertEquals(String.format("Right hand side should return three rows it returned %d", rightHandSide.size()), 3,
+          rightHandSide.size());
+    }
+
+    // Verify that enumerator was called only once for each row.
+    Arrays.asList(mockEnumerableFirstRow, mockEnumerableSecondRow, mockEnumerableThirdRow)
+        .forEach(rowEnumerable -> verify(rowEnumerable, times(1)).enumerator());
+
+    // Test to make sure we created a clone only 3 times.
+    verify(mockBaseEnumerable, times(3)).clone(any());
+  }
+
```

Comment:
Might be worth while testing that a cache size of 2 would change the number of calls to  enumerator and clone.

Reply:
This was really interesting ! So when `compute` on the `Map` is called, it adds the missing `CachingEnumerable` *and then* kicks one out. So you end up cycling whenever the cache size is < batch size recreating each item in the cache on each iteration .